rsconnect/api.py

```diff
@@ -268,6 +268,403 @@ def output_task_log(task_status, last_status, log_callback):
+        return func(*args, **kwargs)
+
+    @console_logged("Validating server...")
+    def validate_server(
+        self,
+        name: str = None,
+        url: str = None,
+        api_key: str = None,
+        insecure: bool = False,
+        cacert: IO = None,
+        api_key_is_required: bool = False,
+        **kwargs
+    ):
+        """
+        Validate that the user gave us enough information to talk to a Connect server.
+
+        :param name: the nickname, if any, specified by the user.
+        :param url: the URL, if any, specified by the user.
+        :param api_key: the API key, if any, specified by the user.
+        :param insecure: a flag noting whether TLS host/validation should be skipped.
+        :param cacert: the file object of a CA certs file containing certificates to use.
+        :param api_key_is_required: a flag that notes whether the API key is required or may
+        be omitted.
+        """
+        url = url or self.connect_server.url
+        api_key = api_key or self.connect_server.api_key
+        insecure = insecure or self.connect_server.insecure
+        api_key_is_required = api_key_is_required or self.get("api_key_is_required", **kwargs)
+        server_store = ServerStore()
+
+        if cacert:
+            ca_data = text_type(cacert.read())
+        else:
+            ca_data = self.connect_server.ca_data
+
+        if name and url:
+            raise RSConnectException("You must specify only one of -n/--name or -s/--server, not both.")
+
+        real_server, api_key, insecure, ca_data, from_store = server_store.resolve(
+            name, url, api_key, insecure, ca_data
+        )
+
+        # This can happen if the user specifies neither --name or --server and there's not
+        # a single default to go with.
+        if not real_server:
+            raise RSConnectException("You must specify one of -n/--name or -s/--server.")
+
+        connect_server = RSConnectServer(real_server, None, insecure, ca_data)
+
+        # If our info came from the command line, make sure the URL really works.
+        if not from_store:
+            self.server_settings
+
+        connect_server.api_key = api_key
+
+        if not connect_server.api_key:
+            if api_key_is_required:
+                raise RSConnectException('An API key must be specified for "%s".' % connect_server.url)
+            return self
+
+        # If our info came from the command line, make sure the key really works.
+        if not from_store:
+            _ = self.verify_api_key()
+
+        self.connect_server = connect_server
+        self.connect = RSConnect(self.connect_server)
+
+        return self
+
+    @console_logged("Making bundle ...")
+    def make_bundle(self, func, *args, **kwargs):
+        path = (
+            self.get("path", **kwargs)
+            or self.get("file", **kwargs)
+            or self.get("file_name", **kwargs)
+            or self.get("directory", **kwargs)
```

Comment:
This is caused by the different upstream bundle functions that take either a file or a folder. Unification needs to happen there first before it can be simplified here.